lib/Backend/IRBuilderAsmJs.cpp

```diff
@@ -1897,23 +1897,36 @@ IRBuilderAsmJs::BuildAsmReg1(Js::OpCodeAsmJs newOpcode, uint32 offset)
 void
 IRBuilderAsmJs::BuildAsmReg1(Js::OpCodeAsmJs newOpcode, uint32 offset, Js::RegSlot dstReg)
 {
-    Assert(newOpcode == Js::OpCodeAsmJs::LdUndef);
+    Assert(newOpcode == Js::OpCodeAsmJs::LdUndef || newOpcode == Js::OpCodeAsmJs::CurrentMemory_Int);
 
-    Js::RegSlot dstRegSlot = GetRegSlotFromVarReg(dstReg);
+    if (newOpcode == Js::OpCodeAsmJs::LdUndef)
+    {
+
+        Js::RegSlot dstRegSlot = GetRegSlotFromVarReg(dstReg);
+        IR::RegOpnd * dstOpnd = BuildDstOpnd(dstRegSlot, TyVar);
 
-    IR::RegOpnd * dstOpnd = BuildDstOpnd(dstRegSlot, TyVar);
+        if (dstOpnd->m_sym->m_isSingleDef)
+        {
+            dstOpnd->m_sym->m_isConst = true;
+            dstOpnd->m_sym->m_isNotInt = true;
+        }
 
-    if (dstOpnd->m_sym->m_isSingleDef)
+        IR::AddrOpnd * addrOpnd = IR::AddrOpnd::New(m_func->GetScriptContext()->GetLibrary()->GetUndefined(), IR::AddrOpndKindDynamicVar, m_func, true);
+        addrOpnd->SetValueType(ValueType::Undefined);
+
+        IR::Instr * instr = IR::Instr::New(Js::OpCode::Ld_A, dstOpnd, addrOpnd, m_func);
+        AddInstr(instr, offset);
+    }
+    else
     {
-        dstOpnd->m_sym->m_isConst = true;
-        dstOpnd->m_sym->m_isNotInt = true;
+        Js::RegSlot dstRegSlot = GetRegSlotFromIntReg(dstReg);
+        IR::RegOpnd * dstOpnd = BuildDstOpnd(dstRegSlot, TyInt32);
+        Assume(m_asmFuncInfo->UsesHeapBuffer());
```

Comment:
I don't think this can be assumed. However, I think we should be able to implement such that this is invariant (e.g. if max > 0, we always allocate at least one page). this way, we can have `if(UsesHeapBuffer) { AddInstr(dst = length >> 16)} else { AddInstr(dst = 0)}`